[SPARK-18967][SCHEDULER] compute locality levels even if delay = 0

squito opened the pull request:
## What changes were proposed in this pull request?

Before this change, with delay scheduling off, spark would effectively
ignore locality preferences for bulk scheduling.  With this change,
locality preferences are used when multiple offers are made
simultaneously.

## How was this patch tested?

Test case added which fails without this change.  All unit tests run via jenkins.

squito committed:
[SPARK-18967][SCHEDULER] compute locality levels even if delay = 0

Before this change, with delay scheduling off, spark would effectively
ignore locality preferences for bulk scheduling.  This ensures that
locality preferences are used when multiple offers are made
simultaneously, and adds a test case for it.

SparkQA commented:
**[Test build #70494 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70494/testReport)** for PR 16376 at commit [`341b73a`](https://github.com/apache/spark/commit/341b73a9255bb159af82b855df1a628caf3ec152).

squito committed:
fix test name

SparkQA commented:
**[Test build #70495 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70495/testReport)** for PR 16376 at commit [`6347464`](https://github.com/apache/spark/commit/6347464d21da175c076dbc7cccc873d3d5fc1e6c).

mridulm commented:
At first glance, this looks like the right change - but I might be missing something.
Also +CC @lirui-intel who made this change.

SparkQA commented:
**[Test build #70494 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70494/testReport)** for PR 16376 at commit [`341b73a`](https://github.com/apache/spark/commit/341b73a9255bb159af82b855df1a628caf3ec152).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70494/
Test FAILed.

vanzin commented:
It'd be nice to check whether the test case attached to SPARK-1937 still applies. Reading the description of that bug, it seems related not only to this but to other things you've been playing with recently.

SparkQA commented:
**[Test build #70495 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70495/testReport)** for PR 16376 at commit [`6347464`](https://github.com/apache/spark/commit/6347464d21da175c076dbc7cccc873d3d5fc1e6c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70495/
Test PASSed.

lirui-intel commented:
The change looks good to me, although I still want to make sure I understand it correctly. Before the change, a locality level is invalid if it has delay=0. The patch changes that and makes such locality level valid too. So during scheduling, it gives us a chance to try the smaller levels first, e.g. PROCESS_LOCAL, NODE_LOCAL, otherwise we'll just schedule the task to any executors available. And meanwhile, since the wait time for the smaller levels are 0, we'll quickly fall back to ANY if a satisfying executor is not available. So that the delay=0 is still enforced. Is this correct?

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:
```diff
@@ -338,7 +338,7 @@ private[spark] class TaskSchedulerImpl private[scheduler](
     }.getOrElse(offers)
 
     // Randomly shuffle offers to avoid always placing tasks on the same set of workers.
```
you can remove this comment now that it's duplicated in the shuffleOffers method description

kayousterhout reviewed (commented):
This looks good.  Can you also update the comment on myLocalityLevels in TaskSetManager.scala to clarify what it's used for?  It currently mentions delay scheduling, which is no longer true after this PR.

Similar to @vanzin, I was a little nervous about this change so tracked down the origin of it and why it was added as part of #892 (for SPARK-1937).  It looks like the bug you've fixed was added towards the end of the review process for #892: it was added in the [2nd to last commit](https://github.com/apache/spark/pull/892/commits/fafd57f483cead8b52a13dc36a1ddb60c7bd3a14) in that PR, and apparently as a result of the June 10th comment beginning with "I've updated...".  That comment implies that this bug was introduced as a last-minute opportunistic performance improvement (and not as part of fixing the bug that PR was addressing).  Specifically, that comment/commit changed TaskSchedulerImpl to iterate only though TSM.myLocalityLevels, rather than through all of the locality levels.  So, you could also fix this issue by un-doing that change, but I think the approach you took here is more intuitive (and results in slightly better performance, since it does avoid iterating over levels for which nothing will happen, which is exactly what the original buggy change was intended to do).

kayousterhout commented:
@lirui-intel yes, that's consistent with my understanding.  The TaskSetManager still checks that it's not going beyond the currently-allowed locality level [here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L424), so locality levels > 0 will still be enforced, and the TaskSchedulerImpl is still checking all of the locality levels in myLocalityLevels each time it has an offer, so it will still schedule non-local tasks when the wait is 0 and no local tasks can be scheduled.  Also, my comment above may be helpful for understanding why this bug was originally introduced.

lirui-intel commented:
@kayousterhout I see. Thanks for the explanations :)

squito committed:
Merge branch 'master' into locality_without_delay

squito committed:
update comments

squito committed:
use manual clock in test; add test for non-preferred scheduling with delay scheduling off

SparkQA commented:
**[Test build #70936 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70936/testReport)** for PR 16376 at commit [`9c64888`](https://github.com/apache/spark/commit/9c648885a410b8c7e55b8a13f88efde8460b9639).

squito committed:
typo

squito commented:
Thanks for the feedback.  I've updated the comment on `myLocalityLevels`.

Also I updated the tests slightly.  To ensure that we're really testing no delay, I updated the tests to use a manual clock which never advances.  And also to address @lirui-intel 's concern, i also added a test that we still schedule at non-preferred locations immediately with delay scheduling off.  (Somewhat obvious in current implementation since we always add `ANY` but I think its a good regression test if nothing else.)

SparkQA commented:
**[Test build #70937 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70937/testReport)** for PR 16376 at commit [`930c2b7`](https://github.com/apache/spark/commit/930c2b7b02360a6ce3f01db5420f3f326daaf282).

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -159,7 +159,12 @@ private[spark] class TaskSetManager(
+   * This allows a performance optimization, of skipping levels that aren't relevant (eg., skip
+   * PROCESS_LOCAL if no tasks could be run PROCESS_LOCAL for the current set of executors).
+   */
   var myLocalityLevels = computeValidLocalityLevels()
```
As I was figuring out the purpose of this for what to put in the comment, I made a couple of observations:

1) For each executor we add or remove, its an O(numExecutors) operation to update the locality levels.  So overall its an O(numExecutors^2) to add a bunch.  Minor on small clusters, but I wonder if this is an issue when you're using dynamic allocation and going up and down to 1000s of executors.  Its all happening with a lock on the `TaskSchedulerImpl` too.

2) Though we recompute valid locality levels as executors come and go, we do *not* as tasks complete.  That's not a problem -- as offers come in, we still go through the right task lists.  But it does make me wonder whether this business of updating the locality levels for the current set of executors is useful, and instead we should just always use all levels.

kayousterhout replied:
(1) does seem like an issue.  I also mostly agree for (2), since the logic of avoiding unnecessarily waiting for delay timeouts is already handled (separately from the myLocalityLevels calculation) [here](https://github.com/squito/spark/blob/930c2b7b02360a6ce3f01db5420f3f326daaf282/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L548).  My only hesitation is that myLocalityLevels does allow avoiding the delay timeout in cases where there are tasks have constraints to run on executors that haven't been granted to the application, so that use case seems like it might merit keeping the code (also, if you agree, can you update the myLocalityLevels comment?).  In any case I'd do this in a separate PR.

kayousterhout reviewed (commented):
LGTM. New test looks great!

SparkQA commented:
**[Test build #70937 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70937/testReport)** for PR 16376 at commit [`930c2b7`](https://github.com/apache/spark/commit/930c2b7b02360a6ce3f01db5420f3f326daaf282).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70937/
Test FAILed.

SparkQA commented:
**[Test build #70936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70936/testReport)** for PR 16376 at commit [`9c64888`](https://github.com/apache/spark/commit/9c648885a410b8c7e55b8a13f88efde8460b9639).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70936/
Test PASSed.

squito commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #71084 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71084/testReport)** for PR 16376 at commit [`930c2b7`](https://github.com/apache/spark/commit/930c2b7b02360a6ce3f01db5420f3f326daaf282).

SparkQA commented:
**[Test build #71084 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71084/testReport)** for PR 16376 at commit [`930c2b7`](https://github.com/apache/spark/commit/930c2b7b02360a6ce3f01db5420f3f326daaf282).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71084/
Test PASSed.

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:
```diff
@@ -54,7 +54,7 @@ import org.apache.spark.util.{AccumulatorV2, ThreadUtils, Utils}
     val sc: SparkContext,
     val maxTaskFailures: Int,
-    blacklistTrackerOpt: Option[BlacklistTracker],
+    val blacklistTrackerOpt: Option[BlacklistTracker],
```
Just noticed that this doesn't seem to be used publicly anywhere -- does it need to be a val? Should it be private val?

squito replied:
oh good point, this change is not necessary (must have been part of another change which I later reverted, sorry)

squito replied:
doh, actually this is used in tests.  I override `createTaskSetManager` to use my manual clock.  Another alternative is to allow the clock to get passed in to the `TaskSCheudlerImpl` constructor, but that ends up being a bigger code change.

I'll make it `private[scheduler] val`, slightly tighter than the implicit `private[spark]` it would be otherwise.

squito committed:
Merge branch 'master' into locality_without_delay

squito committed:
cleanup

SparkQA commented:
**[Test build #71137 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71137/testReport)** for PR 16376 at commit [`c7e7d6e`](https://github.com/apache/spark/commit/c7e7d6e6302d41cc79d2d60da8d9019b28f4dcfb).

SparkQA commented:
**[Test build #71137 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71137/testReport)** for PR 16376 at commit [`c7e7d6e`](https://github.com/apache/spark/commit/c7e7d6e6302d41cc79d2d60da8d9019b28f4dcfb).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71137/
Test FAILed.

squito committed:
fix

SparkQA commented:
**[Test build #71141 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71141/testReport)** for PR 16376 at commit [`e2344d2`](https://github.com/apache/spark/commit/e2344d2449c11d9ecec0f301771ff016725dd9f1).

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:
```diff
@@ -819,4 +819,84 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
+    }
+  }
+
+  test("With delay scheduling off, tasks can be run at any locality level immediately") {
```
Sorry last thing: just realized -- does this test need to first submit a local resource offer?  That makes sure that the local executor is considered alive.  Otherwise, process local won't be in the set of allowed locality levels because of the code here: https://github.com/apache/spark/pull/16376/files#diff-bad3987c83bd22d46416d3dd9d208e76R966, which makes this test somewhat less effective if I understand correctly

squito replied:
yes, you are absolutely right.  I've updated and also added a check to make sure tsm includes lower locality levels.

squito committed:
cleanup

SparkQA commented:
**[Test build #71156 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71156/testReport)** for PR 16376 at commit [`0852742`](https://github.com/apache/spark/commit/08527428138b19b0b064632b3dcfd05524c243ac).

SparkQA commented:
**[Test build #71141 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71141/testReport)** for PR 16376 at commit [`e2344d2`](https://github.com/apache/spark/commit/e2344d2449c11d9ecec0f301771ff016725dd9f1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71141/
Test PASSed.

kayousterhout commented:
LGTM

mridulm commented:
LGTM, thanks @squito !

SparkQA commented:
**[Test build #71156 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71156/testReport)** for PR 16376 at commit [`0852742`](https://github.com/apache/spark/commit/08527428138b19b0b064632b3dcfd05524c243ac).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71156/
Test FAILed.

squito commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #71223 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71223/testReport)** for PR 16376 at commit [`0852742`](https://github.com/apache/spark/commit/08527428138b19b0b064632b3dcfd05524c243ac).

SparkQA commented:
**[Test build #71223 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71223/testReport)** for PR 16376 at commit [`0852742`](https://github.com/apache/spark/commit/08527428138b19b0b064632b3dcfd05524c243ac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71223/
Test PASSed.

kayousterhout commented:
@squito I just noticed this hasn't been merged.  Is this good to go pending tests passing again?

kayousterhout commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #72477 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72477/testReport)** for PR 16376 at commit [`0852742`](https://github.com/apache/spark/commit/08527428138b19b0b064632b3dcfd05524c243ac).

SparkQA commented:
**[Test build #72477 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72477/testReport)** for PR 16376 at commit [`0852742`](https://github.com/apache/spark/commit/08527428138b19b0b064632b3dcfd05524c243ac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72477/
Test PASSed.

squito committed:
style

squito committed:
Merge branch 'master' into locality_without_delay

squito commented:
yes, I think this is ready (I just noticed a couple of minor nits with a fresh read but no real changes)

SparkQA commented:
**[Test build #72484 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72484/testReport)** for PR 16376 at commit [`cf8271e`](https://github.com/apache/spark/commit/cf8271e4c69efa573aabf891d5465736c7c79184).

SparkQA commented:
**[Test build #72484 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72484/testReport)** for PR 16376 at commit [`cf8271e`](https://github.com/apache/spark/commit/cf8271e4c69efa573aabf891d5465736c7c79184).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72484/
Test PASSed.

kayousterhout commented:
Awesome always enthusiastic about fixing minor nits!!  I merged this into master.  I didn't merge it into 2.1 but I don't feel strongly about it.
